docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,276 @@
+
+Contrast that with scenarios for using `DisallowNull`: You use this attribute to specify that an input variable of a nullable type shouldn't be `null`. Consider a property where `null` is the default value, but clients can only set it to a non-null value. Consider the following code:
+
+```csharp
+public string ReviewComment // Comments can be set, but not cleared.
+{
+    get { return _comment;}
+    set
+    {
+        if (value == null) throw new ArgumentNullException(nameof(value), "Cannot set to null");
+        _comment = null;
+    }
+}
+string _comment;
+```
+
+The preceding code is the best way to express your design that the `ReviewComment` could be `null`, but can't be set to `null`. Once this code is nullable aware, you can express this concept more clearly to callers using the <xref:System.Diagnostics.CodeAnalysis.DisallowNull?displayProperty=nameWithType>:
+
+```csharp
+[DisallowNull] 
+public string? ReviewComment // Comments can be added, but not removed.
+{
+    get { return _comment;}
+    set
+    {
+        if (value == null) throw new ArgumentNullException(nameof(value), "Cannot set to null");
+        _comment = null;
+    }
+}
+string? _comment;
+```
+
+In a nullable context, the preceding code warns callers that the `ReviewComment` could be `null`, so it must be checked before access. Furthermore, it warns callers that, even though it could be `null`, callers shouldn't explicitly set it to `null`. The `DisallowNull` attribute also specifies a *pre-condition*, it does not affect the `get` accessor. You should choose to use the `DisallowNull` attribute when you observe these characteristics about:
+
+1. The variable could be `null` in core scenarios, often when first instantiated.
+1. The variable shouldn't be explicitly set to `null`.
+
+These situations are common in code that was originally *null oblivious*. It may be that object properties are set in two distinct initialization operations. It may be that some properties are set only after some asynchronous work has completed.
+
+The `AllowNull` and `DisallowNull` attributes enable you to specify that preconditions on variables may not match the nullable annotations on those variables. These provide more detail about the characteristics of your API. This additional information helps callers use your API correctly.
+
+## Specify post-conditions: `MaybeNull` and `NotNull`
+
+Suppose you have a method with the following signature:
+
+```csharp
+public Customer FindCustomer(string lastName, string firstName)
+```
+
+You've likely written a method like this to return `null` when the name sought wasn't found. The `null` clearly indicates that the record wasn't found. In this example, you'd likely change the return type from `Customer` to `Customer?`. Instead, suppose it was a generic method like the following code:
```

Comment:
Why would I likely have changed it to `Customer?`? 

Also, I don't like the changing of the example below to `Find` instead of `FindCustomer`. It makes it feel like the introduction of `FindCustomer` was pointless. Maybe continue with `FindCustomer` and use the `return: MaybeNull` on it?

Reply:
The problem is that `T?` doesn't work. That's covered later. I added text to allude to that fact.